zilliqa/src/node.rs

```diff
@@ -263,26 +263,65 @@ impl Node {
         self.consensus.head_block().header.number
     }
 
-    pub fn resolve_block_number(&self, number: BlockNumberOrTag) -> u64 {
-        // TODO(#863): This doesn't respect two-chain finalization.
-        match number {
-            BlockNumberOrTag::Latest => self.get_chain_tip(),
-            BlockNumberOrTag::Finalized => self.get_chain_tip().saturating_sub(2),
-            BlockNumberOrTag::Safe => self.get_chain_tip().saturating_sub(2),
+    pub fn resolve_block_number(&self, block_number: BlockNumberOrTag) -> u64 {
+        match block_number {
+            BlockNumberOrTag::Number(n) => n,
             BlockNumberOrTag::Earliest => 0,
+            BlockNumberOrTag::Latest => self.get_chain_tip(),
             BlockNumberOrTag::Pending => self.get_chain_tip(),
```

Comment:
`latest` and `pending` have different semantics:
- `latest` = the most recent block that has been proposed and received
- `pending` = the block that we would be propose as next, containing transactions from the mempool